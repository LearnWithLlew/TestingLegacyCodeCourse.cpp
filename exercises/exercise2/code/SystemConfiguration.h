#pragma once

#include <string>
#include <sstream>

int rpm = 4000;
double dt = 0.01;
double x = 0;
double y = 0;
double z = 0;
double x_mph = 1;
double y_mph = 2;
double z_mph = 3;
bool gravity = true;
double y_torque = 3.2;

void initialiseSystem()
{
    if ( gravity )
    {
        double dx = x_mph / ( 60 * 60 * rpm );
        double dy = y_mph / ( 60 * 60 * rpm );
        for ( int i = 1; i <= rpm / 60; ++i )
        {
            x += dt * dx;
            y += dt * dy;
            z += dt * z_mph / ( 60 * 60 * rpm );
        }
    }
    else
    {
        double dx = x_mph / ( 60 * 60 * rpm );
        double dy = y_mph / ( 60 * 60 * rpm );
        for ( int i = 1; i <= rpm / 60; ++i )
        {
            x += dt * dx;
            y += dt * dy;
        }
    }
}
