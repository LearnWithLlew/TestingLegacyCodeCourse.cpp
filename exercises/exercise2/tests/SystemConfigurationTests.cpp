#include <catch2/catch.hpp>
#include "ApprovalTests.hpp"

#include "SystemConfiguration.h"

#include <iostream>
#include <sstream>

using namespace ApprovalTests;
#define PRINT(x) ss << #x " " << x << std::endl;


std::string printGlobals() {

	std::stringstream ss;
	PRINT(rpm);
	PRINT(dt);
	PRINT(x);
	PRINT(y);
	PRINT(z);
	PRINT(x_mph);
	PRINT(y_mph);
	PRINT(z_mph);
	PRINT(gravity);
	PRINT(y_torque);

	return ss.str();
}

void setAllGlobals(int _rpm = 4000,
double _dt = 0.01,
double _x = 0,
double _y = 0,
double _z = 0,
double _x_mph = 1,
double _y_mph = 2,
double _z_mph = 3,
bool _gravity = true,
double _y_torque = 3.2
) {
	rpm = _rpm;
	dt = _dt;
	x = _x;
	y = _y;
	z = _z;
	x_mph = _x_mph;
	y_mph = _y_mph;
	z_mph = _z_mph;
	gravity = _gravity;
	y_torque = _y_torque;
}

std::string GenerateCode()
{
	std::stringstream ss;
	ss << "setAllGlobals("
		<< rpm << ", "
		<< dt << ", "
		<< x << ", "
		<< y << ", "
		<< z << ", "
		<< x_mph << ", "
		<< y_mph << ", "
		<< z_mph << ", "
		<< gravity << ", "
		<< y_torque << ");";
	return ss.str();
}

TEST_CASE("Test System Configuration 1")
{
	std::stringstream ss;
	setAllGlobals(4000, 0.01);
	ss << "Initial State: " << std::endl;
	ss << printGlobals();
	initialiseSystem();

	ss << "Modified State: " << std::endl;
	ss << printGlobals();
	Approvals::verify(ss.str());
}


TEST_CASE("Test System Configuration 2")
{
	std::stringstream ss;
	setAllGlobals(4000, 0.01, 4.58333e-08,9.16667e-08,1.375e-07);
	ss << "Initial State: " << std::endl;
	ss << printGlobals();
	initialiseSystem();

	ss << "Modified State: " << std::endl;
	ss << printGlobals();
	Approvals::verify(ss.str());
}



TEST_CASE("generate code")
{
	setAllGlobals(4000, 0.01, 9.16666e-08, 1.83333e-07, 2.75e-07, 1, 2, 3, 1, 3.2);
	Approvals::verify(GenerateCode());
}
