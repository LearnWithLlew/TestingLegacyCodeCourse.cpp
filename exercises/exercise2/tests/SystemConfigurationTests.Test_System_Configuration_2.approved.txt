Initial State: 
rpm 4000
dt 0.01
x 4.58333e-08
y 9.16667e-08
z 1.375e-07
x_mph 1
y_mph 2
z_mph 3
gravity 1
y_torque 3.2
Modified State: 
rpm 4000
dt 0.01
x 9.16666e-08
y 1.83333e-07
z 2.75e-07
x_mph 1
y_mph 2
z_mph 3
gravity 1
y_torque 3.2

